### PR TITLE
[MU4] Fix #10517 KeySig refactor

### LINKED ISSUE #10517: [MU4 Task] KeySym refactor
**Task description**
Remove `KeySym` members  `pos` and `spos`, and instead of them, add
```
int line;
double xPos; // in staff spatium units
```
Why?
`pos` and `spos` are "visual" members (spos.y depends on line distance), so same Key Signature had different Key Symbols spos values in different staff settings. 
In custom key signatures, spos was (incorrectly) used as "logical" member instead, which caused visual bug #10497 in staffs with altered line distance.

`line` and `xPos` are "logical" members, which clean up code logic and brings more flexibility to Key Signatures.
(For example it opens up opportunities to adapt custom key signatures per clef, or to transpose them in future)

### PR DESCRIPTION
**Pull request by sammik**
Resolves: #10517 
Also Resolves: #10497, Closes: #10516, Closes: #10543 

Remove `KeySym` members `pos` and `spos`, and instead of them, add
```
int line;
double xPos; // in staff spatium units
```
### Why?
`pos` and `spos` are "visual" members (spos.y depends on line distance), so same Key Signature had different Key Symbols spos values in different staff settings.
In custom key signatures, spos was (incorrectly) used as "logical" member instead, which caused visual bug https://github.com/musescore/MuseScore/issues/10497 in staffs with altered line distance.

`line` and `xPos` are "logical" members, which clean up code logic and brings more flexibility to Key Signatures.
(For example it opens up opportunities to adapt custom key signatures per clef, or to transpose them in future)

### Save / Import Export
KeySym is saved into file like this:
```
<KeySym>
  <sym>accidentalSharp</sym>
  <pos xPos="0" line="0"/>
  </KeySym>

```
Before this change, it was `<pos x="0" y="0"/>`. Code is compatible with old files, they are opened correctly.

Import and export to musicxml works too.

--
I used `qreal stepDistance = staff() ? staff()->staffTypeForElement(this)->lineDistance().val() * 0.5 : 0.5;` in few places, so I also changed `qreal stepDistance = staff() ? staff()->lineDistance(tick()) * 0.5 : 0.5;` in one place to it, in case of consistency.

--
tested on Linux (Ubuntu Studio 21.10)

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

**Comment by sammik** [comment]
Failing visual test is actually good, as this also fixes custom key sig visual bug.

**Comment by Jojo-Schmitz** [comment]
so a `vtests` label is in order I guess

**Comment by sammik** [comment]
> so a `vtests` label is in order I guess

You mean, I shoul tick "I created the test" option too?

**Comment by sammik** [comment]
> No, someone with sufficient priviledge should add the `vtests` label

Ah, OK, thanks for explanation.

**Review by cbjeukendrup — COMMENTED** [review]
Looks like you've done a good job! Just a few minor comments.

File: src/engraving/libmscore/key.h [hunk 1]
```diff
@@ -77,8 +77,8 @@ static inline Key operator-=(Key& a, const Key& b) { return a = Key(static_cast<
 
 struct KeySym {
     SymId sym;
-    mu::PointF spos;       // position in spatium units
-    mu::PointF pos;        // actual pixel position on screen (set by layout)
+    int line;       //relative line position (first staffline: line == 0, first gap: line == 1, ...)
```
Comment (Jojo-Schmitz) [inline src/engraving/libmscore/key.h:80]: I'd rather see a space between the `//` and the comment's text

File: src/engraving/libmscore/keysig.cpp [hunk 2]
```diff
@@ -437,7 +445,10 @@ void KeySig::read(XmlReader& e)
                     }
                     ks.sym = id;
                 } else if (t == "pos") {
-                    ks.spos = e.readPoint();
+                    int idx = _sig.keySymbols().size();
+                    ks.xPos = e.doubleAttribute("xPos", e.doubleAttribute("x", idx));  //if xPos not there, use x, or accidetal index
+                    ks.line = e.intAttribute("line", static_cast<int>(e.doubleAttribute("y", 2) * 2)); //if line not there, use y, or middle line
```
Comment (Jojo-Schmitz) [inline src/engraving/libmscore/keysig.cpp:450]: here too

Comment (cbjeukendrup) [reply src/engraving/libmscore/keysig.cpp:450]: And maybe put these comments above the line they belong to, because this way the lines are too long.

**Comment by sammik** [comment]
Should be done. Thank You both.

**Review by cbjeukendrup — APPROVED** [review]
Looks good to me, thanks!

**Comment by cbjeukendrup** [comment]
Looks like a good optimization :)
I still think `_spatium = spatium()` makes sense, because the `_spatium` variable is used more than once. By creating a variable, the spatium only needs to be calculated/retrieved once.

**Comment by sammik** [comment]
Replaced by #11280

**Pull request closed without merging**